parachain/src/primitives.rs

```diff
@@ -230,4 +225,10 @@ pub struct ValidationResult {
 	pub head_data: HeadData,
 	/// An update to the validation code that should be scheduled in the relay chain.
 	pub new_validation_code: Option<ValidationCode>,
+	/// Upward messages send by the Parachain.
+	pub upward_messages: Vec<UpwardMessage>,
+	/// Number of downward messages that were processed by the Parachain.
+	///
+	/// It is expected that the Parachain processes them from first to last.
+	pub processed_downward_messages: u32,
```

Comment:
Why not require that all downward messages are processed? Or that all downward messages submitted at/before block X are processed?

Reply:
I see the problem that if we require that the Parachain processes all the downward messages, it could to overload the Parachain. 
Let's say the Parachain only can process X downward messages per block at maximum, because after that all the block weight is used. This would mean the Parachain can not execute any other extrinsics. 

In the end this is some sort of back pressure, which let's the Parachain decide on how it processes the messages. Maybe we need to ensure in the future that we at least process one downward message per relay chain block, but this whole implementation is only be meant to have something to show.

Reply:
This is also a problem faced on the full XCMP protocol which is also addressed in a similar way.

However, the way we address it there is for the collator to decide to include all messages up to block X. So you can't include some message from block X without including all of them.

Doing things this way will make the migration to XCMP simpler, even though it will require adding some extra information to the queue.

Reply:
Agree. A parachain should be able to accept all downward messages, even if they just end up getting queued for processing in a later block. One thing we might consider is allowing a parachain to flag that it cannot accept any further downward messages for the time being.

Reply:
The middle ground that we came to for XCMP is that parachains progress a watermark, which is monotonically increasing, and indicates the block number that all messages sent at or before have been processed. Then we can set rules about how the watermark must advance (minimum of `k` blocks each time, to the most recent block number, at least half of the distance between last watermark and relay chain head number, etc.).

Doing the same here seems sensible, and then we can set watermark processing rules that ensure paras process messages faster than they receive them.